ouroboros-network/src/Ouroboros/Network/BlockFetch/ClientState.hs

```diff
@@ -200,22 +200,34 @@ initialPeerFetchInFlight =
 
 addHeadersInFlight :: HasHeader header
                    => (header -> SizeInBytes)
-                   -> FetchRequest header
-                   -> Int
-                   -- ^ the number of new requests; this might differ from the
-                   -- length of 'FetchRequest' by one, since when we write to
-                   -- @TFetchRequestVar@ we combine 'FetchRequest's using their
-                   -- semigroup instance, which might merge two chain fragments.
+                   -> Maybe (FetchRequest header) -- ^ Existing request if any.
+                   -> FetchRequest header         -- ^ The extra request.
                    -> PeerFetchInFlight header
                    -> PeerFetchInFlight header
-addHeadersInFlight blockFetchSize (FetchRequest fragments) newRequests inflight =
+addHeadersInFlight blockFetchSize existing (FetchRequest fragments) inflight =
+
+    -- This assertion checks the pre-condition 'addNewFetchRequest' that all
+    -- requested blocks are new. This is true irrespective of fetch-request
+    -- command merging.
     assert (and [ blockPoint header `Set.notMember` peerFetchBlocksInFlight inflight
                 | fragment <- fragments
                 , header   <- CF.toOldestFirst fragment ]) $
+
     PeerFetchInFlight {
+
+      -- We have the ability to merge new fetch requests with existing (as-yet
+      -- unacknowledged) fetch requests when the old and new turn out to be for
+      -- contigious ranges. This means that we end up sending fewer requests
+      -- but for bigger ranges. It also makes tracking the number of requests
+      -- in-flight a bit more subtle. We have to look at the /difference/ in
+      -- the number of fragments (corresponding to one request each) for the
+      -- existing (if any) and new request.
       peerFetchReqsInFlight   = peerFetchReqsInFlight inflight
-                              + fromIntegral newRequests,
+                              + fromIntegral (length fragments)
+                              - maybe 0 (length . fetchRequestFragments) existing,
```

Comment:
`length . fetchRequests`, will produce an `Int`, but `peerFetchReqsInFlight :: Word`.

Reply:
I did say I'd not even compiled it :-)  (was having troubles with ghc versions). Thanks, will get those sorted and update the PR.

Reply:
That's the only problem CI found.